charger/fritzdect.go

```diff
@@ -0,0 +1,301 @@
+	default:
+		return false, fmt.Errorf("state unknown result: %d", status.State)
+	}
+}
+
+// Enable implements the Charger.Enable interface
+func (c *FritzDECT) Enable(enable bool) (err error) {
+	var status fritzdectStatusResponse
+	if enable {
+		status, err = c.apiUpdate("SetSwitchOn")
+	} else {
+		status, err = c.apiUpdate("SetSwitchOff")
+	}
+	if err != nil {
+		return err
+	}
+	if enable && validFritzBoxResponse(status) && status.State == 0 {
+		return fmt.Errorf("wasn't able to switchOn: %d", status.State)
+	}
+	if !enable && validFritzBoxResponse(status) && status.State == 1 {
+		return fmt.Errorf("wasn't able to switchOff: %d", status.State)
+	}
+	return nil
+}
+
+// MaxCurrent implements the Charger.MaxCurrent interface (Dummy function)
+func (c *FritzDECT) MaxCurrent(current int64) error {
+	// FritzDECT switch has no option to set MaxCurrent
+	return nil
+}
+
+// CurrentPower implements the Meter interface.
+func (c *FritzDECT) CurrentPower() (float64, error) {
+	status, err := c.apiStatus()
+	if status.Power < c.standbypower {
+		return 0, err
+	}
+	return status.Power, err
+}
+
+// ChargedEnergy implements the ChargeRater interface
+func (c *FritzDECT) ChargedEnergy() (float64, error) {
+	status, err := c.apiStatus()
+	energy := status.Energy / 1000
+	return energy, err
+}
+
+// Fritzbox helpers (based on ideas of https://github.com/rsdk/ahago)
+
+//getFritzBoxSessionID fetches a session-id based on the username and password in the connection struct
+func (c *FritzDECT) getFritzBoxSessionID() error {
+	uri := fmt.Sprintf("%s/login_sid.lua", c.uri)
+	parameters := make(map[string]string)
+	type Result struct {
+		SID       string
+		Challenge string
+		BlockTime string
+	}
+	body, err := sendFritzBoxRequest(uri, parameters)
+	if err != nil {
+		return err
+	}
+	v := Result{SID: "none", Challenge: "none", BlockTime: "none"}
+	err = xml.Unmarshal(body, &v)
+	if err != nil {
+		return err
+	}
+	if v.SID == "0000000000000000" {
+		parameters["username"] = c.user
+		parameters["response"] = createFbChallengeResponse(v.Challenge, c.password)
+		body, err = sendFritzBoxRequest(uri, parameters)
+		if err != nil {
+			return err
+		}
+		err = xml.Unmarshal(body, &v)
+		if err != nil {
+			return err
+		}
+	}
+	c.sid = v.SID
+	return nil
+}
+
+// sendFritzBoxRequest sends a HTTP Request to the FritzBox based on the given URL and returns the answer as a bytearray
+func sendFritzBoxRequest(uri string, parametersIn map[string]string) ([]byte, error) {
+	var URL *url.URL
+	URL, err := url.Parse(uri)
+	if err != nil {
+		return nil, err
+	}
+	parameters := url.Values{}
+	for key, value := range parametersIn {
+		parameters.Add(key, value)
+	}
+	URL.RawQuery = parameters.Encode()
+	resp, err := http.Get(URL.String())
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != 200 {
+		return nil, errors.New("sendFritzBoxRequest status code != 200")
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
```

Comment:
Return ioutil.Readall....

usw: das geht in Summe deutlich einfacher mit weniger Code ;)